feature: added 'comments' functionality

inwakeofquake opened the pull request:
Added extra feature giving admin and users an option to leave comments to events.

inwakeofquake committed:
feature: added classes for 'comments' functionality - entity, dto, controllers, repo, service
fix: added transactional annotations to all services
add: link to github repo in README.md
add: basic postman tests

inwakeofquake committed:
fix: CommentDto DATE link fix

inwakeofquake committed:
feature: exceptions structure and logic severely simplified

inwakeofquake committed:
fix: remove call to 'save' method in EventServiceImpl update methods
fix: remove redundant datePattern fields in multiple classes

inwakeofquake committed:
fix: change logging style in CommentServiceImpl
fix: add isBlank checks for String fields in EventServiceImpl methods

inwakeofquake committed:
fix: add isBlank checks for String fields in CompilationServiceImpl methods
fix: removed 'final' modifier from String fields in UserIncomeDto

inwakeofquake committed:
fix: removed @RequiredArgsConstructor in UserIncomeDto, added constructor

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/dto/comment/NewCommentDto.java:
```diff
@@ -0,0 +1,17 @@
+@NoArgsConstructor
+public class NewCommentDto {
+    @NotBlank
+    @Size
```
Так какой размер?)

inwakeofquake replied:
Добавил max 255, как в схеме

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/dto/user/UserIncomeDto.java:
```diff
@@ -2,19 +2,22 @@
-    private final String email;
+    private String email;
+
+    public UserIncomeDto(String name, String email) {
```
Почему решил руками прописать конструктор?

inwakeofquake replied:
Потому что ты просил убрать ключевое слово final у полей, а в таком случае ломбок уже не может на выдачу сделать конструктор с обеими полями для требований логики приложения, потому что подцепляет только финальные поля.

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/entity/Comment.java:
```diff
@@ -0,0 +1,29 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id", nullable = false)
+    private Long id;
```
В чем смысл прописывать nullable = false для генерируемого id?)

inwakeofquake replied:
Убрал

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/category/CategoryServiceImpl.java:
```diff
@@ -19,6 +18,7 @@
 @Service
 @RequiredArgsConstructor
+@Transactional(readOnly = true)
 public class CategoryServiceImpl implements CategoryService {
```
Логирование для сервиса

inwakeofquake replied:
Добавил

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/category/CategoryServiceImpl.java:
```diff
@@ -59,9 +59,10 @@ public void deleteCategory(Long catId) {
+            throw new ConflictException("Can't update category, name: already used");
         }
         category.setName(categoryDto.getName());
         return categoryMapper.toCategoryDto(categoryRepository.save(category));
```
Вызов метода save() при обновлении не требуется

inwakeofquake replied:
Убрал

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+    public CommentDto createComment(NewCommentDto newCommentDto, Long userId, Long eventId) {
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new NotFoundException(String.format(
+                        "Can't create comment, user with id=%s not found", userId)));
```
Здесь и далее логирование при неудачных попытках

inwakeofquake replied:
Добавил

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+            throw new ConflictException("Can't delete comment, must be owner");
+        }
+        oldComment.setText(newCommentDto.getText());
+        Comment savedComment = commentsRepository.save(oldComment);
```
Вызов метода save() не требуется

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+
+    @Override
+    @Transactional
+    public CommentDto createComment(NewCommentDto newCommentDto, Long userId, Long eventId) {
```
Логирование при положительном результате

inwakeofquake replied:
Добавил

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+        Comment oldComment = commentsRepository.findById(commentId).orElseThrow(() -> new NotFoundException(
+                "Can't update comment, comment not found"));
+        oldComment.setText(newCommentDto.getText());
+        Comment savedComment = commentsRepository.save(oldComment);
```
И тут

TarasovSam replied:
Не поправил

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+            LocalDateTime createEnd,
+            Integer from,
+            Integer size) {
+        userRepository.findById(userId);
```
Что дает эта строка?

inwakeofquake replied:
Удалил

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+
+    @Override
+    public List<CommentDto> getCommentsByEventIdByAdmin(Long eventId, Integer from, Integer size) {
+        eventRepository.findById(eventId).orElseThrow(() -> new NotFoundException(String.format(
```
Что дает эта строка?

inwakeofquake replied:
Переделал логику метода

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,183 @@
+    @Override
+    @Transactional
+    public void deleteCommentByAdmin(Long commentId) {
+        commentsRepository.deleteById(commentId);
```
А если commentId не существует?

inwakeofquake replied:
Добавил проверочный блок и логирование

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/compilation/CompilationServiceImpl.java:
```diff
@@ -28,6 +28,7 @@
 @Service
 @RequiredArgsConstructor
 @Slf4j
```
Не везде прописано логирование при отрицательных и положительных результатах

inwakeofquake replied:
Добавил

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/event/EventServiceImpl.java:
```diff
@@ -36,30 +38,29 @@
 @Service
 @RequiredArgsConstructor
+@Transactional(readOnly = true)
 public class EventServiceImpl implements EventService {
```
В сервисе нет логирования

inwakeofquake replied:
Добавил. Заодно перегруппировал методы

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/request/RequestServiceImpl.java:
```diff
@@ -21,6 +22,7 @@
 @Service
 @RequiredArgsConstructor
+@Transactional(readOnly = true)
 public class RequestServiceImpl implements RequestService {
```
Логирование

inwakeofquake replied:
Добавил

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/request/RequestServiceImpl.java:
```diff
@@ -84,12 +86,12 @@ public RequestStatusUpdateResult updateRequests(
-            throw new ParticipantLimitException("exceeding the limit of participants");
+            throw new ConflictException("Limit of participants exceeded");
         }
 
```
На 107 строке вызов save()

inwakeofquake replied:
Убрал

inwakeofquake replied:
Убрал

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/statistics/StatisticsServiceImpl.java:
```diff
@@ -19,8 +19,7 @@
 public class StatisticsServiceImpl implements StatisticsService {
```
Логирование
Транзакционность

inwakeofquake replied:
Ты уверен, что здесь есть методы, которым нужно добавить транзакционность? Ни один метод не меняет здесь состояние бд.

inwakeofquake replied:
Логирование добавил

TarasovSam replied:
Ну так и укажи для всего класса readOnly=true :)

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/user/UserServiceImpl.java:
```diff
@@ -17,6 +17,7 @@
 @Slf4j
 @RequiredArgsConstructor(onConstructor_ = @Autowired)
+@Transactional(readOnly = true)
 public class UserServiceImpl implements UserService {
```
Ниже написать не могу
В методе удаления нет проверки на то, что id существует

inwakeofquake replied:
Добавил

TarasovSam commented on main-service/src/main/java/ru/practicum/main_service/service/user/UserServiceImpl.java:
```diff
@@ -17,6 +17,7 @@
 @Service
 @Slf4j
```
Нет логирования при отрицательных результатах

inwakeofquake replied:
Добавил

TarasovSam reviewed (commented):
Андрей, привет!
Проверил твое задание, все очень неплохо) Оставил комментарии с вопросами и рекомендациями (многие из них повторяются)
Жду твою работу на следующую итерацию!) Если будут вопросы, пиши в пачку)

inwakeofquake committed:
fix: add size limit to NewCommentDto
fix: removed nullable false in Comment
feature: added logging to CategoryServiceImpl
feature: added logging to CommentServiceImpl
fix: updated methods in CommentServiceImpl
feature: added some logging to CompilationServiceImpl

inwakeofquake committed:
feature: added logging to EventServiceImpl, rearranged methods
fix: cut redundant lines of code in CompilationServiceImpl

inwakeofquake committed:
fix: removed redundant calls to 'save' method in RequestServiceImpl
feature: added logging to RequestServiceImpl
fix: added exists check to deleteUser method in UserServiceImpl
feature: added logging to StatisticsServiceImpl
fix: removed redundant call to 'save' method in CategoryServiceImpl
feature: added logging to UserServiceImpl

inwakeofquake committed:
fix: removed redundant call to 'save method in updateCommentByUser method in CommentServiceImpl'

TarasovSam reviewed (commented):
Андрей, поздравляю с успешной сдачей финального спринта дипломного проекта! Я рад, что ты справился со всеми трудностями и дошел до конца! Желаю успехов в дальнейшем обучении и во всех начинаниях!)
